node/block_exchange.go

```diff
@@ -0,0 +1,233 @@
+package node
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"github.com/celestiaorg/go-header"
+	goheaderp2p "github.com/celestiaorg/go-header/p2p"
+	goheaderstore "github.com/celestiaorg/go-header/store"
+	goheadersync "github.com/celestiaorg/go-header/sync"
+	ds "github.com/ipfs/go-datastore"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/net/conngater"
+	"github.com/tendermint/tendermint/libs/log"
+	tmtypes "github.com/tendermint/tendermint/types"
+	"go.uber.org/multierr"
+
+	"github.com/rollkit/rollkit/config"
+	"github.com/rollkit/rollkit/p2p"
+	"github.com/rollkit/rollkit/types"
+)
+
+// P2P Exchange Service for block that implements the go-header interface.
+// Contains a block store where synced blocks are stored.
+// Uses the go-header library for handling all P2P logic.
+type BlockExchangeService struct {
+	conf         config.NodeConfig
+	genesis      *tmtypes.GenesisDoc
+	p2p          *p2p.Client
+	ex           *goheaderp2p.Exchange[*types.Block]
+	syncer       *goheadersync.Syncer[*types.Block]
+	sub          *goheaderp2p.Subscriber[*types.Block]
+	p2pServer    *goheaderp2p.ExchangeServer[*types.Block]
+	blockStore   *goheaderstore.Store[*types.Block]
+	syncerStatus *SyncerStatus
```

Comment:
[non-blocking]
I think we can improve how we are handling the syncer status. First, as a nit, I would group the related fields to make it more explicit.
```suggestion
	sub          *goheaderp2p.Subscriber[*types.Block]
	p2pServer    *goheaderp2p.ExchangeServer[*types.Block]
	blockStore   *goheaderstore.Store[*types.Block]
	
	syncer       *goheadersync.Syncer[*types.Block]
	syncerStatus *SyncerStatus
```

Second, I think it makes since to have a helper method of the `BlockExchangeService` called `StartSyncer`:

```go
func (bes *BlockExchangeService) StartSyncer() error {
  bes.syncerStatus.mu.Lock()
  defer bes.syncerStatus.mu.Unlock()
  if bes.syncerStatus.started {
    return nil
  }
  err := bes.syncer.Start()
  if err != nil {
    return err
  }
  bes.syncerStatus.started = true
  return nil
}
```